hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java

```diff
@@ -3900,4 +3900,26 @@ private void updateLayoutVersionInDB(OMLayoutVersionManager lvm,
   private BucketLayout getBucketLayout() {
     return BucketLayout.DEFAULT;
   }
+
+  /**
+   * Get supported Bucket Layouts based on upgrade state.
+   * Before finalization, the method returns an empty list. This leads to the
+   * new client's OzoneClientAdapter and CreateBucketHandler receiving an empty
+   * list - which causes them to continue with LEGACY layout.
+   * After finalization, the method returns a list with the new layouts.
+   *
+   * @return Supported Bucket Layouts.
+   */
+  public List<BucketLayout> getSupportedBucketLayouts() {
+    // Check if we are in Pre-Finalize upgrade state.
+    if (versionManager.getUpgradeState() ==
+        UpgradeFinalizer.Status.FINALIZATION_REQUIRED) {
```

Comment:
Add FSO as a layout feature in `OMLayoutFeature`. Then you can call `OzoneManager#getVersionManager#isAllowed(OMLayoutFeature.FSO)` to determine whether FSO has been finalized or not.